File: app/src/main/java/org/wikipedia/readinglist/database/ReadingList.java
package org.wikipedia.readinglist.database;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.R;
import org.wikipedia.WikipediaApp;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class ReadingList implements Serializable {
    public static final int SORT_BY_NAME_ASC = 0;
    public static final int SORT_BY_NAME_DESC = 1;
    public static final int SORT_BY_RECENT_ASC = 2;
    public static final int SORT_BY_RECENT_DESC = 3;
    public static final ReadingListTable DATABASE_TABLE = new ReadingListTable();

    private List<ReadingListPage> pages = new ArrayList<>();

    private long id;
    @NonNull private String title;
    @Nullable private String description;
    private long mtime;
    private long atime;
    private long sizeBytes;
    private boolean dirty = true;
    private long remoteId;

    @Nullable private transient String accentAndCaseInvariantTitle;

    protected ReadingList(@NonNull String title, @Nullable String description) {
        this.title = title;
        this.description = description;
        long now = System.currentTimeMillis();
        mtime = now;
        atime = now;
    }

    public List<ReadingListPage> pages() {
        return pages;
    }

    public int numPagesOffline() {
        int count = 0;
        for (ReadingListPage page : pages) {
            if (page.offline() && page.status() == ReadingListPage.STATUS_SAVED) {
                count++;
            }
        }
        return count;
    }

    public boolean isDefault() {
        return TextUtils.isEmpty(title);
    }

    public long id() {
        return id;
    }
    public void id(long id) {
        this.id = id;
    }

    @NonNull public String title() {
        return isDefault() ? WikipediaApp.getInstance().getString(R.string.default_reading_list_name) : title;
    }
    public void title(@NonNull String title) {
        this.title = title;
    }
    public String dbTitle() {
        return title;
    }

    @NonNull public String accentAndCaseInvariantTitle() {
        if (accentAndCaseInvariantTitle == null) {
            accentAndCaseInvariantTitle = StringUtils.stripAccents(title).toLowerCase();
        }
        return accentAndCaseInvariantTitle;
    }

    @Nullable public String description() {
        return description;
    }
    public void description(@Nullable String description) {
        this.description = description;
    }

    public long getMTime() {
        return mtime;
    }
    public void setMTime(long mtime) {
        this.mtime = mtime;
    }

    public long atime() {
        return atime;
    }
    public void atime(long atime) {
        this.atime = atime;
    }

    public void touch() {
        atime = System.currentTimeMillis();
    }

    public long sizeBytes() {
        long bytes = 0;
        for (ReadingListPage page : pages) {
            bytes += page.offline() ? page.sizeBytes() : 0;
        }
        return bytes;
    }
    public void sizeBytes(long sizeBytes) {
        this.sizeBytes = sizeBytes;
    }

    public boolean dirty() {
        return dirty;
    }
    public void dirty(boolean dirty) {
        this.dirty = dirty;
    }

    public long remoteId() {
        return remoteId;
    }
    public void remoteId(long remoteId) {
        this.remoteId = remoteId;
    }

    public static void sort(ReadingList list, int sortMode) {
        switch (sortMode) {
            case SORT_BY_NAME_ASC:
                Collections.sort(list.pages(), Comparator.comparing(ReadingListPage::accentAndCaseInvariantTitle));
                break;
            case SORT_BY_NAME_DESC:
                Collections.sort(list.pages(), Comparator.comparing(ReadingListPage::accentAndCaseInvariantTitle).reversed());
                break;
            case SORT_BY_RECENT_ASC:
                Collections.sort(list.pages(), Comparator.comparingLong(ReadingListPage::getMTime));
                break;
            case SORT_BY_RECENT_DESC:
                Collections.sort(list.pages(), Comparator.comparingLong(ReadingListPage::getMTime).reversed());
                break;
            default:
                break;
        }
    }

    public static void sort(List<ReadingList> lists, int sortMode) {
        switch (sortMode) {
            case SORT_BY_NAME_ASC:
                Collections.sort(lists, Comparator.comparing(ReadingList::accentAndCaseInvariantTitle));
                break;
            case SORT_BY_NAME_DESC:
                Collections.sort(lists, Comparator.comparing(ReadingList::accentAndCaseInvariantTitle).reversed());
                break;
            case SORT_BY_RECENT_ASC:
                Collections.sort(lists, Comparator.comparingLong(ReadingList::getMTime));
                break;
            case SORT_BY_RECENT_DESC:
                Collections.sort(lists, Comparator.comparingLong(ReadingList::getMTime).reversed());
                break;
            default:
                break;
        }
        // make the Default list sticky on top, regardless of sorting.
        ReadingList defaultList = null;
        for (ReadingList list : lists) {
            if (list.isDefault()) {
                defaultList = list;
                break;
            }
        }
        if (defaultList != null) {
            lists.remove(defaultList);
            lists.add(0, defaultList);
        }
    }

    public static void sortGenericList(List<Object> lists, int sortMode) {
        switch (sortMode) {
            case SORT_BY_NAME_ASC:
                Collections.sort(lists, (lhs, rhs) -> {
                    if (lhs instanceof ReadingList && rhs instanceof ReadingList) {
                        return ((ReadingList) lhs).accentAndCaseInvariantTitle().compareTo(((ReadingList) rhs).accentAndCaseInvariantTitle());
                    } else {
                        return 0;
                    }
                });
                break;
            case SORT_BY_NAME_DESC:
                Collections.sort(lists, (lhs, rhs) -> {
                    if (lhs instanceof ReadingList && rhs instanceof ReadingList) {
                        return ((ReadingList) rhs).accentAndCaseInvariantTitle().compareTo(((ReadingList) lhs).accentAndCaseInvariantTitle());
                    } else {
                        return 0;
                    }
                });
                break;
            case SORT_BY_RECENT_ASC:
                Collections.sort(lists, (lhs, rhs) -> {
                    if (lhs instanceof ReadingList && rhs instanceof ReadingList) {
                        return Long.compare(((ReadingList) rhs).getMTime(), ((ReadingList) lhs).getMTime());
                    } else {
                        return 0;
                    }
                });
                break;
            case SORT_BY_RECENT_DESC:
                Collections.sort(lists, (lhs, rhs) -> {
                    if (lhs instanceof ReadingList && rhs instanceof ReadingList) {
                        return Long.compare(((ReadingList) lhs).getMTime(), ((ReadingList) rhs).getMTime());
                    } else {
                        return 0;
                    }
                });
                break;
            default:
                break;
        }
        // make the Default list sticky on top, regardless of sorting.
        ReadingList defaultList = null;
        for (Object list : lists) {
            if (list instanceof ReadingList && ((ReadingList) list).isDefault()) {
                defaultList = (ReadingList) list;
                break;
            }
        }
        if (defaultList != null) {
            lists.remove(defaultList);
            lists.add(0, defaultList);
        }
    }
}

File: app/src/main/java/org/wikipedia/readinglist/database/ReadingList.kt
package org.wikipedia.readinglist.database

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import org.apache.commons.lang3.StringUtils
import org.wikipedia.R
import org.wikipedia.util.L10nUtil
import java.io.Serializable

// TODO: create default reading list upon initial DB creation.

@Entity
class ReadingList(
    var listTitle: String,
    var description: String?,
    var mtime: Long = System.currentTimeMillis(),
    var atime: Long = mtime,
    @PrimaryKey(autoGenerate = true) var id: Long = 0,
    var sizeBytes: Long = 0,
    var dirty: Boolean = true,
    var remoteId: Long = 0
) : Serializable {

    @Ignore
    val pages = mutableListOf<ReadingListPage>()

    @delegate:Transient
    val accentInvariantTitle: String by lazy {
        StringUtils.stripAccents(title)
    }

    @Transient
    var selected = false

    var title
        get() = listTitle.ifEmpty { L10nUtil.getString(R.string.default_reading_list_name) }
        set(value) { listTitle = value }

    val isDefault
        get() = title == L10nUtil.getString(R.string.default_reading_list_name)

    val numPagesOffline
        get() = pages.count { it.offline && it.status == ReadingListPage.STATUS_SAVED }

    val sizeBytesFromPages
        get() = pages.sumOf { if (it.offline) it.sizeBytes else 0 }

    fun touch() {
        atime = System.currentTimeMillis()
    }

    fun compareTo(other: Any): Boolean {
        return (other is ReadingList &&
                id == other.id &&
                pages.size == other.pages.size &&
                numPagesOffline == other.numPagesOffline &&
                title == other.title &&
                description == other.description)
    }

    companion object {
        const val SORT_BY_NAME_ASC = 0
        const val SORT_BY_NAME_DESC = 1
        const val SORT_BY_RECENT_ASC = 2
        const val SORT_BY_RECENT_DESC = 3

        fun sort(list: ReadingList, sortMode: Int) {
            when (sortMode) {
                SORT_BY_NAME_ASC -> list.pages.sortWith { lhs: ReadingListPage, rhs: ReadingListPage -> lhs.accentInvariantTitle.compareTo(rhs.accentInvariantTitle, true) }
                SORT_BY_NAME_DESC -> list.pages.sortWith { lhs: ReadingListPage, rhs: ReadingListPage -> rhs.accentInvariantTitle.compareTo(lhs.accentInvariantTitle, true) }
                SORT_BY_RECENT_ASC -> list.pages.sortWith { lhs: ReadingListPage, rhs: ReadingListPage -> lhs.mtime.compareTo(rhs.mtime) }
                SORT_BY_RECENT_DESC -> list.pages.sortWith { lhs: ReadingListPage, rhs: ReadingListPage -> rhs.mtime.compareTo(lhs.mtime) }
            }
        }

        fun sort(lists: MutableList<ReadingList>, sortMode: Int) {
            when (sortMode) {
                SORT_BY_NAME_ASC -> lists.sortWith { lhs: ReadingList, rhs: ReadingList -> lhs.accentInvariantTitle.compareTo(rhs.accentInvariantTitle, true) }
                SORT_BY_NAME_DESC -> lists.sortWith { lhs: ReadingList, rhs: ReadingList -> rhs.accentInvariantTitle.compareTo(lhs.accentInvariantTitle, true) }
                SORT_BY_RECENT_ASC -> lists.sortWith { lhs: ReadingList, rhs: ReadingList -> rhs.mtime.compareTo(lhs.mtime) }
                SORT_BY_RECENT_DESC -> lists.sortWith { lhs: ReadingList, rhs: ReadingList -> lhs.mtime.compareTo(rhs.mtime) }
            }
            // make the Default list sticky on top, regardless of sorting.
            lists.firstOrNull { it.isDefault }?.let {
                lists.remove(it)
                lists.add(0, it)
            }
        }

        fun sortGenericList(lists: MutableList<Any>, sortMode: Int) {
            when (sortMode) {
                SORT_BY_NAME_ASC -> lists.sortWith { lhs: Any?, rhs: Any? ->
                    if (lhs is ReadingList && rhs is ReadingList) {
                        lhs.accentInvariantTitle.compareTo(rhs.accentInvariantTitle, true)
                    } else {
                        0
                    }
                }
                SORT_BY_NAME_DESC -> lists.sortWith { lhs: Any?, rhs: Any? ->
                    if (lhs is ReadingList && rhs is ReadingList) {
                        rhs.accentInvariantTitle.compareTo(lhs.accentInvariantTitle, true)
                    } else {
                        0
                    }
                }
                SORT_BY_RECENT_ASC -> lists.sortWith { lhs: Any?, rhs: Any? ->
                    if (lhs is ReadingList && rhs is ReadingList) {
                        rhs.mtime.compareTo(lhs.mtime)
                    } else {
                        0
                    }
                }
                SORT_BY_RECENT_DESC -> lists.sortWith { lhs: Any?, rhs: Any? ->
                    if (lhs is ReadingList && rhs is ReadingList) {
                        lhs.mtime.compareTo(rhs.mtime)
                    } else {
                        0
                    }
                }
            }

            // make the Default list sticky on top, regardless of sorting.
            lists.firstOrNull { it is ReadingList && it.isDefault }?.let {
                lists.remove(it)
                lists.add(0, it)
            }
        }
    }
}
